### PR TITLE
Expose a non dummy derivative index lowering

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/ModelingToolkit.jl/issues/2513

### CODE REVIEW

**Comment by baggepinnen** [comment]
Is this supposed to be user facing?

**Comment by ChrisRackauckas** [comment]
To an extent, because it has different trade-offs in the system that is generated. However, the non-DD version seems to be a bit buggy so I'm not sure if it's ready.

**Comment by ChrisRackauckas** [comment]
MWE:

```julia
using ModelingToolkit, OrdinaryDiffEq, Test
using ModelingToolkit: t_nounits as t, D_nounits as D
@parameters g
@variables x(t) [state_priority = 10] y(t) λ(t)

eqs = [
       D(D(x)) ~ λ * x
       D(D(y)) ~ λ * y - g
       x^2 + y^2 ~ 1
      ]
@named pend = ODESystem(eqs,t)
pend = complete(pend)
ss = structural_simplify(pend; dummy_derivative = false)

unknowns(ss)
equations(ss)

#=
julia> unknowns(ss)
7-element Vector{SymbolicUtils.BasicSymbolic{Real}}:
 x(t)
 y(t)
 xˍtt(t)
 yˍtt(t)
 λ(t)
 xˍt(t)
 yˍt(t)

julia> equations(ss)
5-element Vector{Equation}:
 Differential(t)(x(t)) ~ xˍt(t)
 Differential(t)(y(t)) ~ yˍt(t)
 0 ~ -xˍtt(t) + x(t)*λ(t)
 0 ~ -g - yˍtt(t) + y(t)*λ(t)
 0 ~ 1 - (x(t)^2) - (y(t)^2)
=#

@named pend = ODESystem(eqs,t)
ss = complete(dae_index_lowering(expand_connections(pend)))

#=
julia> equations(ss)
3-element Vector{Equation}:
 Differential(t)(Differential(t)(x(t))) ~ x(t)*λ(t)
 Differential(t)(Differential(t)(y(t))) ~ -g + y(t)*λ(t)
 0 ~ -2(Differential(t)(y(t))^2) - 2(Differential(t)(x(t))^2) - 2(x(t)^2)*λ(t) - 2y(t)*(-g + y(t)*λ(t))
 =#

 ss = complete(ode_order_lowering(dae_index_lowering(expand_connections(pend))))

 #=
 julia> equations(ss)
5-element Vector{Equation}:
 Differential(t)(xˍt(t)) ~ x(t)*λ(t)
 Differential(t)(yˍt(t)) ~ -g + y(t)*λ(t)
 Differential(t)(x(t)) ~ xˍt(t)
 Differential(t)(y(t)) ~ yˍt(t)
 0 ~ -2(Differential(t)(y(t))^2) - 2(Differential(t)(x(t))^2) - 2(x(t)^2)*λ(t) - 2y(t)*(-g + y(t)*λ(t))
 =#
 ```

It seems just doing `dummy_derivative=false` isn't what I expected. I do think as a debugging tool we should expose the last to the user somehow in a documented way. @YingboMa how would you prefer that be done?

**Comment by YingboMa** [comment]
We can make `structural_simplify(sys, dummy_derivative = false)` behave like `structural_simplify(dae_index_lowering(expand_connections(sys)))`.

**Comment by ChrisRackauckas** [comment]
The manual order lowering is required? At least from my tests there. Is there a way to make pantelides do that part?

**Comment by YingboMa** [comment]
I am not sure if `ode_order_lowering(dae_index_lowering(expand_connections(pend)))` always return a valid semi-implicit ODE in general.

**Comment by YingboMa** [comment]
> The manual order lowering is required? At least from my tests there. Is there a way to make pantelides do that part?

I think the only reliable way of lowering to first order is `structural_simplify`.

**Comment by ChrisRackauckas** [comment]
That's what I tried first, but the structural_simplify way (this PR) gives an invalid system (the code in this PR), so something went wrong 😅 .

File: src/systems/systemstructure.jl [hunk 1]
```diff
@@ -628,7 +629,7 @@ function _structural_simplify!(state::TearingState, io; simplify = false,
     if check_consistency
         ModelingToolkit.check_consistency(state, orig_inputs)
     end
-    if fully_determined
+    if fully_determined && dummy_derivative
         sys = ModelingToolkit.dummy_derivative(sys, state; simplify, mm, check_consistency)
     else
         sys = ModelingToolkit.tearing(sys, state; simplify, mm, check_consistency)
```
Comment (YingboMa) [inline src/systems/systemstructure.jl:635]: else you need to call `dae_index_lowering` before tearing I think.

Comment (ChrisRackauckas) [reply src/systems/systemstructure.jl:635]: only if it's an ODESystem I presume?

Comment (YingboMa) [reply src/systems/systemstructure.jl:635]: Only call `dae_index_lowering` if `fully_determined`, of course.

Comment (YingboMa) [reply src/systems/systemstructure.jl:635]: I think `fully_determined` is more general than `ODESystem`.

Comment (ChrisRackauckas) [reply src/systems/systemstructure.jl:635]:
```
7-element Vector{SymbolicUtils.BasicSymbolic{Real}}:
 x(t)
 y(t)
 xˍtt(t)
 yˍtt(t)
 λ(t)
 xˍt(t)
 yˍt(t)

5-element Vector{Equation}:
 Differential(t)(x(t)) ~ xˍt(t)
 Differential(t)(y(t)) ~ yˍt(t)
 0 ~ -xˍtt(t) + x(t)*λ(t)
 0 ~ -g - yˍtt(t) + y(t)*λ(t)
 0 ~ 1 - (x(t)^2) - (y(t)^2)

```

it's giving back weird systems from that, and still doing DD?

Comment (YingboMa) [reply src/systems/systemstructure.jl:635]: That's a bit unexpected. What does `tearing(dae_index_lowering(expand_connections(sys)))` on the REPL do?

Comment (ChrisRackauckas) [reply src/systems/systemstructure.jl:635]: Same:

```julia
using ModelingToolkit, OrdinaryDiffEq, Test
using ModelingToolkit: t_nounits as t, D_nounits as D
@parameters g
@variables x(t) [state_priority = 10] y(t) λ(t)

eqs = [
       D(D(x)) ~ λ * x
       D(D(y)) ~ λ * y - g
       x^2 + y^2 ~ 1
      ]
@named pend = ODESystem(eqs,t)
sys = tearing(dae_index_lowering(expand_connections(pend)))
```

```julia
7-element Vector{SymbolicUtils.BasicSymbolic{Real}}:
 y(t)
 x(t)
 xˍtt(t)
 yˍtt(t)
 λ(t)
 xˍt(t)
 yˍt(t)

julia> equations(sys)
5-element Vector{Equation}:
 Differential(t)(y(t)) ~ yˍt(t)
 Differential(t)(x(t)) ~ xˍt(t)
 0 ~ -xˍtt(t) + x(t)*λ(t)
 0 ~ -g - yˍtt(t) + y(t)*λ(t)
 0 ~ -2(xˍt(t)^2) - 2(yˍt(t)^2) - 2(x(t)^2)*λ(t) - 2y(t)*(-g + y(t)*λ(t))
```

Comment (YingboMa) [reply src/systems/systemstructure.jl:635]: I know this is silly, but it works
```julia
julia> sys = dummy_derivative(dae_index_lowering(expand_connections(pend)));

julia> equations(sys)
5-element Vector{Equation}:
 Differential(t)(xˍt(t)) ~ x(t)*λ(t)
 Differential(t)(yˍt(t)) ~ -g + y(t)*λ(t)
 Differential(t)(y(t)) ~ yˍt(t)
 Differential(t)(x(t)) ~ xˍt(t)
 0 ~ -2(xˍt(t)^2) - 2(yˍt(t)^2) - 2(x(t)^2)*λ(t) - 2y(t)*(-g + y(t)*λ(t))
```
I guess `tearing` doesn't work well with high order systems.

**Comment by ChrisRackauckas** [comment]
Silly is fine if it's correct 😅. It's a bit odd to implement in practice because I assume it needs to be after the validity checks, but then there's state in the tearing state so I reconstruct it. Let me know if this state reconstruction should have another step in there.

**Comment by YingboMa** [comment]
I don’t think it’s valid to keep the state. I can take a look tomorrow.

**Comment by ChrisRackauckas** [comment]
I don't keep the state, I generate a new one.

**Comment by YingboMa** [comment]
Ah, I meant mm.

**Review by YingboMa — APPROVED** [review]

**Pull request merged**